### ISSUE TITLE
Remove use of `bridgesAllowedLength` from `HomeBridgeErcToNative.setBlockRewardContract` function

### ISSUE THREAD

**Issue by varasev** [poanetwork/poa-bridge-contracts#194]
Since the `BlockReward` contract in the [`posdao-contracts`](https://github.com/poanetwork/posdao-contracts) repo doesn't have `bridgesAllowed` and `bridgesAllowedLength` getters anymore, we should remove the use of `bridgesAllowedLength` from the [`HomeBridgeErcToNative.setBlockRewardContract`](https://github.com/poanetwork/poa-bridge-contracts/blob/bbb97a63c900f03a902d0e82358abac3b294e4d9/contracts/upgradeable_contracts/erc20_to_native/HomeBridgeErcToNative.sol#L119) function.

**Comment by akolotov** [comment]
It makes sense to use https://github.com/poanetwork/posdao-contracts/blob/master/contracts/BlockRewardAuRa.sol source for reference of the contract that will be used in the POSDAO environment.

**Comment by akolotov** [comment]
The suggestion is to use
```solidity
(IBlockReward(_blockReward).ercToNativeBridgesAllowed().length != 0)
```
instead of 
```solidity
(IBlockReward(_blockReward).bridgesAllowedLength() != 0)
```

**Comment by patitonar** [comment]
When compiling for `spuriousDragon` we get the following error

```
TypeError: Member "length" not found or not visible after argument-dependent lookup in inaccessible dynamic type
        require(_blockReward != address(0) && isContract(_blockReward) && (IBlockReward(_blockReward).ercToNativeBridgesAllowed().length != 0));
```
The problem is that `ercToNativeBridgesAllowed` returns `address[] memory` https://github.com/poanetwork/posdao-contracts/blob/6e510e1d721fc931ff65014e5c7cba626e1693a8/contracts/abstracts/BlockRewardBase.sol#L172

One option could be to call `DELEGATORS_ALIQUOT` constant variable that returns `uint256`, `bridgesAllowedLength` was a constant variable too so I don't think there will be any issue https://github.com/poanetwork/posdao-contracts/blob/6e510e1d721fc931ff65014e5c7cba626e1693a8/contracts/abstracts/BlockRewardBase.sol#L21

What do you think @akolotov ?

**Comment by varasev** [comment]
For backward compatibility with the current xDai network, I'd suggest not to use the checking `(IBlockReward(_blockReward).ercToNativeBridgesAllowed().length != 0)` at all.

> One option could be to call DELEGATORS_ALIQUOT constant variable that returns uint256

The `DELEGATORS_ALIQUOT` is not related to the number of bridges allowed inside the BlockReward contract.

**Comment by akolotov** [comment]
@varasev it was a security audit issue to add this check. This is needed to not allow to set the block reward contract before it is configured in the block reward contract.

**Comment by patitonar** [comment]
Right, here I found the security audit comment of that moment https://hackmd.io/@fOpdbBK7RgOo7RaZrlW3yw/Sk6P7Uqtm?type=view#HomeBridgeErcToNative  We need to call any method to know that it is the block reward contract and not any other contract. `DELEGATORS_ALIQUOT` could work for the new contracts, the only issue I see now is compatibility with the current xDai network, but I don't think there is a method/public variable that could be use that is present in both, new and old, set of contracts

**Comment by varasev** [comment]
Ok, then maybe make it like

```solidity
interface IBlockReward {
    function addExtraReceiver(uint256 _amount, address _receiver) external;
    function mintedTotally() external view returns (uint256);
    function mintedTotallyByBridge(address _bridge) external view returns(uint256);
    function bridgesAllowedLength() external view returns(uint256);
    function ercToNativeBridgesAllowed() external view returns(address[] memory);
    function addBridgeTokenFeeReceivers(uint256 _amount) external;
    function addBridgeNativeFeeReceivers(uint256 _amount) external;
}

...

function setBlockRewardContract(address _blockReward) public onlyOwner {
    bool isOldBlockRewardContract = IBlockReward(_blockReward).bridgesAllowedLength() != 0;
    address[] memory bridgesAllowed = IBlockReward(_blockReward).ercToNativeBridgesAllowed();
    bool isNewBlockRewardContract = bridgesAllowed.length != 0;
    require(_blockReward != address(0) && isContract(_blockReward) && (isOldBlockRewardContract || isNewBlockRewardContract));
    addressStorage[keccak256(abi.encodePacked("blockRewardContract"))] = _blockReward;
}
```

?

**Comment by akolotov** [comment]
@varasev we will still have an issue since cannot pass arrays through memory on `spuriousDragon`. Can we introduce some identifier in the block reward contract like
```
function isRewardContract() public pure returns(bytes4 _data) {
    return bytes4(keccak256(abi.encodePacked("block-reward")));
}
```
For me, it is a good practice to have such identification. In the bridge contracts we have:
https://github.com/poanetwork/poa-bridge-contracts/blob/bbb97a63c900f03a902d0e82358abac3b294e4d9/contracts/upgradeable_contracts/native_to_erc20/HomeBridgeNativeToErc.sol#L88-L90

If not, we need to work this issue around.

@patitonar what if we try to use approach similar to 
https://github.com/poanetwork/poa-bridge-contracts/blob/bbb97a63c900f03a902d0e82358abac3b294e4d9/contracts/ERC677BridgeToken.sol#L70-L75

We will try to call `bridgesAllowedLength` by this approach. If the method exists `call` will return `true`. If `call` returns `false` we will call `validatorSetContract` by similar manner.

E.g. 
```solidity
bool isBlockRewardContract = false;
if IBlockReward(_blockReward).call(abi.encodeWithSignature("bridgesAllowedLength()" {
    isBlockRewardContract = IBlockReward(_blockReward).bridgesAllowedLength() != 0
}
else {
    if IBlockReward(_blockReward).call(abi.encodeWithSignature("validatorSetContract()" {
        isBlockRewardContract = IBlockReward(_blockReward).validatorSetContract() != address(0)
    }
}

**Comment by varasev** [comment]
> we will still have an issue since cannot pass arrays through memory on spuriousDragon

I tried to compile the code from [my above comment](https://github.com/poanetwork/poa-bridge-contracts/issues/194#issuecomment-503259268) with `remix` (Solidity v0.4.24 was selected) but didn't see errors.

Could you try too?

**Comment by varasev** [comment]
> we will still have an issue since cannot pass arrays through memory on spuriousDragon.

Ah, yes, for `spuriousDragon` I see the same error.

> Can we introduce some identifier in the block reward contract like

Ok, I'll add one.

**Comment by varasev** [comment]
Done: https://github.com/poanetwork/posdao-contracts/commit/299b80a9f4b2acf5d5a41cc20e4bfbed3315c7c7

**Comment by akolotov** [comment]
> Done: [poanetwork/posdao-contracts@299b80a](https://github.com/poanetwork/posdao-contracts/commit/299b80a9f4b2acf5d5a41cc20e4bfbed3315c7c7)

Thanks a lot!!!

**Comment by akolotov** [comment]
Addressed in [TokenBridge contracts 2.3.3](https://github.com/poanetwork/poa-bridge-contracts/releases/tag/2.3.3)
